src/class/audio/audio_device.c

```diff
@@ -523,195 +527,88 @@ static bool audiod_tx_done_cb(uint8_t rhport, audiod_interface_t* audio, uint16_
+  // We encode directly into IN EP's FIFO - abort if previous transfer not complete
   TU_VERIFY(!usbd_edpt_busy(rhport, audio->ep_in));
 
   // Determine amount of samples
-  uint16_t const nEndpointSampleCapacity = CFG_TUD_AUDIO_EPSIZE_IN / CFG_TUD_AUDIO_N_CHANNELS_TX / CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_TX;
-  uint16_t nSamplesPerChannelToSend = tu_fifo_count(&audio->tx_ff[0]) / CFG_TUD_AUDIO_TX_ITEMSIZE;
-  uint16_t nBytesToSend;
+  uint16_t const nEndpointSampleCapacity = CFG_TUD_AUDIO_EP_IN_SW_BUFFER_SIZE / CFG_TUD_AUDIO_N_CHANNELS_TX / CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_TX;
+  uint16_t nSamplesPerChannelToSend = tu_fifo_count(&audio->tx_ff[0]); 	// We first look for the minimum number of bytes and afterwards convert it to sample size
   uint8_t cntChannel;
 
   for (cntChannel = 1; cntChannel < CFG_TUD_AUDIO_N_CHANNELS_TX; cntChannel++)
   {
     uint16_t const count = tu_fifo_count(&audio->tx_ff[cntChannel]);
-    if (count / CFG_TUD_AUDIO_TX_ITEMSIZE < nSamplesPerChannelToSend)
+    if (count < nSamplesPerChannelToSend)
     {
-      nSamplesPerChannelToSend = count * CFG_TUD_AUDIO_TX_ITEMSIZE;
+      nSamplesPerChannelToSend = count;
     }
   }
 
+  // Convert to sample size
+  nSamplesPerChannelToSend = nSamplesPerChannelToSend / CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_TX;
+
   // Check if there is enough
-  if (nSamplesPerChannelToSend == 0)
-  {
-    audio->epin_buf_cnt = 0;
-    return true;
-  }
+  if (nSamplesPerChannelToSend == 0)    return true;
 
   // Limit to maximum sample number - THIS IS A POSSIBLE ERROR SOURCE IF TOO MANY SAMPLE WOULD NEED TO BE SENT BUT CAN NOT!
   nSamplesPerChannelToSend = tu_min16(nSamplesPerChannelToSend, nEndpointSampleCapacity);
-  nBytesToSend = nSamplesPerChannelToSend * CFG_TUD_AUDIO_N_CHANNELS_TX * CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_TX;
 
   // Encode
   uint16_t cntSample;
-  uint8_t * pBuff = audio->epin_buf;
-#if CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_TX == 1
-  uint8_t sample;
-#elif CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_TX == 2
-  uint16_t sample;
-#else
-  uint32_t sample;
-#endif
 
-  // TODO: Big endianess handling
   for (cntSample = 0; cntSample < nSamplesPerChannelToSend; cntSample++)
   {
     for (cntChannel = 0; cntChannel < CFG_TUD_AUDIO_N_CHANNELS_TX; cntChannel++)
     {
-      // Get sample from buffer
-      tu_fifo_read_n(&audio->tx_ff[cntChannel], &sample, CFG_TUD_AUDIO_TX_ITEMSIZE);
-
-      // Put it into EP's buffer - Let alignment problems be handled by memcpy
-      memcpy(pBuff, &sample, CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_TX);
-
-      // Advance pointer
-      pBuff += CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_TX;
-    }
-  }
-
-  audio->epin_buf_cnt = nBytesToSend;
-
-  return true;
-}
-
+      // If 8, 16, or 32 bit values are to be copied
+#if CFG_TUD_AUDIO_N_BYTES_PER_SAMPLE_TX == CFG_TUD_AUDIO_TX_ITEMSIZE
+      tu_fifo_read_n_into_other_fifo(&audio->tx_ff[cntChannel], &audio->ep_in_ff, 0, CFG_TUD_AUDIO_TX_ITEMSIZE);
```

Comment:
Ok, I need help. I thought we want to get rid of the middleman EP Buffer between driver fifo and stm32 USB FIFO. But we still have an middleman EP FIFO. Sorry, the modification is a bit overwhelmed for me to fully understand at once.

Reply:
This part regards to PCM encoding where you have e.g. 4 channels which are to be interleaved into one stream. For this you need an additional buffer! This is an outdated version right now, in the most recent one only one layer of FIFOs is used since two FIFOs in a consecutive manner are not necessary.